### DISCUSSION TITLE
v5 authentication array overrides with environment variables

### DISCUSSION THREAD

**Discussion by atlantis** [emqx/emqx#8320] [Q&A] [answered]
Hi there! So on the new version 5.0.0, I have authentication configured like so:

```
authentication = [
  {
    backend = "built_in_database"
    mechanism = "password_based"
    password_hash_algorithm {name = "plain", salt_position = "suffix"}
    user_id_type = "username"
  },
  {
    backend = "http"
    body {password = "${password}", username = "${username}"}
    connect_timeout = "5s"
    enable_pipelining = 100
    headers {"content-type" = "application/json"}
    mechanism = "password_based"
    method = "post"
    pool_size = 8
    request_timeout = "5s"
    ssl {enable = false, verify = "verify_none"}
    url = "http://fake.url"  #change this using ENV variables
  }
]
```

And then I'm trying to override the url with an environment variable... so I tried: 

`EMQX_AUTHENTICATION__2__URL: http://app:3000`

But when emqx starts it still has http://fake.url as the value. Is there another syntax I can use to accomplish overriding the URL? Thanks!

**Comment by zmstone** [comment]
Where is it configured? In emqx.conf or cluster-override.conf?

**Reply by atlantis** [reply]
Right now it's in emqx.conf but i'm happy to put it in either...

**Reply by zmstone** [reply]
Hi. It’s likely a bug if override doesn’t work for emqx.conf

please make sure that data/cluster-override.conf doesn’t have the  value .
The config override order is 

1. Emqx.conf + environment var
1. Cluster-override.conf
1. Local-override.conf

**Reply by atlantis** [reply]
Yeah there's no data/config/cluster-override.conf yet because i'm doing a fresh rebuild with no persistence. And I checked etc/emqx.conf and it still has the fake url as well. Just wondered if there was a new syntax or something for overriding arrays that I'm missing...

**Reply by zmstone** [reply]
No this is the array override syntax (like 4.x)
I’ll have a look at it later.

**Reply by zmstone** [reply]
Btw, override happens at runtime while loading the config, but not by replacing values in emqx.conf (like in 4.2)

**Comment by zmstone** [comment, marked as answer]
it should be
`EMQX_AUTHORIZATION__SOURCES__2__URL`
the example config has `authorization=[]` which is confusing and should be fixed.

a full config example for `authorization` as below.
```
authorization {
  deny_action: ignore
  no_match: allow
  sources: [
    {
      body {password = "${password}", username = "${username}"}
      connect_timeout = "5s"
      enable_pipelining = 100
      headers {"content-type" = "application/json"}
      method = "post"
      pool_size = 8
      request_timeout = "5s"
      ssl {enable = false, verify = "verify_none"}
      type = "http"
      url = "http://127.0.0.1:8080"
    },
    {
      type: file
      path: "etc/acl.conf"
    }
  ]
}
```

edit: please note the http auth source in above example is positioned at `1` but not `2`.

**Reply by atlantis** [reply]
Huh i'm actually trying to override authentication, not authorization - does that have the "sources" child as well?

**Reply by zmstone** [reply]
oops, this seems to be a bug in env override handing. 
From my quick investigation, it has something to do with the array type being positioned at root level.

**Reply by zmstone** [reply]
found the bug, it's not the array and root thing, but something else which is easy to fix.
will be fixed in 5.0.1.

**Reply by zmstone** [reply]
https://github.com/emqx/emqx/pull/8323
